doc/dev/dialects.rst

```diff
@@ -0,0 +1,417 @@
+  associated documentation for dialects developed with the MLIR project or contributed upstream.
+- `defining MLIR dialects <https://mlir.llvm.org/docs/DefiningDialects/>`_: Documentation describing
+  how to define MLIR dialects and all its components.
+- `debugging tips <https://mlir.llvm.org/getting_started/Debugging/>`_: Useful reference for any
+  kind of MLIR development, especially the debugging flags for `mlir-opt` like tools.
+
+The remainder of this document will guide you through, in concrete terms, how to add a new dialect
+to Catalyst.
+
+Catalyst dialect structure
+==========================
+
+A strong design principle in the MLIR project is the avoidance of boilerplate and redundant code,
+favoring a single source of truth to specify IR components. This is achieved via one or more
+descriptive languages based on the `TableGen <https://llvm.org/docs/TableGen/index.html>`_ format
+known from LLVM. Consequently, dialects, as well as many operations, types, attributes, and other
+IR objects, are specified in two parts:
+
+- a *declarative* part: The main way to define new IR objects via a simple language called
+  `Operation Definition Specification (ODS) <https://mlir.llvm.org/docs/DefiningDialects/Operations/>`_.
+  ODS specifications live in TableGen (``.td``) files, which will automatically generate both C++
+  header and implementation files via the build system.
+
+In Catalyst, TableGen files for dialects can be found in the public include directory, under
+`catalyst/mlir/include/\<DialectName\>/IR <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/include/Quantum/IR>`_.
+
+- a *C++* part: Additional hand-written logic, and most transformations, are defined directly in C++
+  source files. An example might be a storage class for a data type, or the implementation of
+  interface methods for an operation declared in ODS.
+  Note that while the usage of ODS is strongly recommended, full control over the underlying
+  objects can always be achieved by directly working with the relevant C++ classes.
+
+In Catalyst, hand-written implementations for classes & methods can be found in the source
+directory, under `catalyst/mlir/lib\/<DialectName\>/IR <https://github.com/PennyLaneAI/catalyst/tree/main/mlir/lib/Quantum/IR>`_.
+
+An alternative tutorial on creating a basic MLIR dialect can also be found in the official docs, see
+`Creating a Dialect <https://mlir.llvm.org/docs/Tutorials/CreatingADialect/>`_.
+
+
+Creating a new dialect
+======================
+
+Let's see how we can add a new dialect to Catalyst, using OpenQASM as an example! Note that all
+auto-generated files and build artifacts mentioned in the guide will only appear at the end of
+the `Using the dialect <#using-the-dialect>`_ section when the ``make dialects`` command is run.
+
+Start by creating a new TableGen file, located at ``mlir/include/OpenQASM/OpenQASM.td``, with the
+following content:
+
+.. code-block:: tablegen
```

Comment:
coming soon! https://github.com/pygments/pygments/pull/2751

Reply:
Yeah saw that 😅 Still no sign of MLIR though :(